pex/cli/commands/lockfile/updater.py

```diff
@@ -0,0 +1,205 @@
+from pex.sorted_tuple import SortedTuple
+from pex.third_party.pkg_resources import Requirement
+from pex.typing import TYPE_CHECKING
+from pex.util import named_temporary_file
+
+if TYPE_CHECKING:
+    import attr  # vendor:skip
+    from typing import Dict, Iterable, Iterator, Mapping, Optional, Union
+else:
+    from pex.third_party import attr
+
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(frozen=True)
+class VersionUpdate(object):
+    original = attr.ib()  # type: Version
+    updated = attr.ib()  # type: Version
+
+
+@attr.s(frozen=True)
+class LockUpdate(object):
+    updated_resolve = attr.ib()  # type: LockedResolve
+    updates = attr.ib()  # type: Mapping[ProjectName, Optional[VersionUpdate]]
+
+
+@attr.s(frozen=True)
+class LockUpdater(object):
+    """Updates a lockfile in whole or in part.
+
+    A lock updater with no updates specified just updates the whole lock file with the latest
+    available distributions that satisfy the locked resolve criteria.
+
+    More interestingly, a lock updater with a set of updates will attempt to alter just the
+    distributions for those updates. The update can be just a plain project name in which case the
+    latest compatible distribution for that project will be used. The update can also be a
+    full-fledged constraint with a specifier, in which case a distribution matching the constraint
+    and still satisfying the rest of the locked resolve will be searched for. This facility allows
+    for targeted upgrades and downgrades of individual projects in a lock that maintain the
+    integrity of the locked resolve.
+
+    When a set of updates is specified, the updater ensures all other project distributions remain
+    unchanged.
+    """
+
+    @classmethod
+    def create(
+        cls,
+        requirements,  # type: Iterable[Requirement]
+        constraints,  # type: Iterable[Requirement]
+        updates,  # type: Iterable[Requirement]
+        lock_configuration,  # type: LockConfiguration
+        pip_configuration,  # type: PipConfiguration
+    ):
+        original_requirements = tuple(str(requirement) for requirement in requirements)
+
+        original_constraints = {
+            ProjectName(constraint.project_name): constraint for constraint in constraints
+        }  # type: Mapping[ProjectName, Requirement]
+
+        update_constraints_by_project_name = {}  # type: Dict[ProjectName, Requirement]
+        for update in updates:
+            constraint_req = Requirement.parse(update)
+            project_name = ProjectName(constraint_req.project_name)
+            original_constraint = original_constraints.get(project_name)
+            if original_constraint:
+                logger.warning(
+                    "Over-riding original constraint {original} with {override}.".format(
+                        original=original_constraint, override=constraint_req
+                    )
+                )
```

Comment:
Should this explode? If you were using a lockfile, it would imply that your constraints file was not being used for that usecase, and instead would likely be a hand-written thing used to pin down your transitive deps: overwriting it here would be strange.

Alternatively, should it try to merge the requirements, and fail if it can't?

Reply:
> Should this explode? If you were using a lockfile, it would imply that your constraints file was not being used for that usecase, and instead would likely be a hand-written thing used to pin down your transitive deps: overwriting it here would be strange.

It doesn't seem strange to me:

1. You use a constraints file and create a lock on some day.
2. Later, you learn one requirement you over-zealously had pinned in the constraints file has a security issue and by upgrading or downgrading you can avoid the security issue and still maintain compatibility.

So you use `-p` to over-ride the constraint. Top actually pass in a new, edited constraints file means re-running the whole lock and you don't want to do that.

What is strange is that the constraints recorded in the lock file are not edited by the over-ride. I'll fix that.